File: backend/app/models/agent_chat.py
"""Persisted recipe-agent chats.

A chat belongs to a single user inside a household. Messages keep the full
conversation including assistant tool calls and tool responses so that the
context can be replayed when the user sends the next message.
"""

from __future__ import annotations

import enum
from datetime import UTC, datetime
from typing import TYPE_CHECKING, Any, Self, cast

from sqlalchemy import func, select
from sqlalchemy.orm import Mapped

from app import db


def _to_utc_ms(value: datetime | None) -> int | None:
    """Serialize a (possibly naive) datetime as UTC milliseconds since epoch.

    The rest of the API serializes datetimes the same way via
    ``KitchenOwlJSONProvider`` -- we mirror it explicitly here for fields
    we have to format ourselves so the frontend can always treat them as
    timezone-aware UTC instants and convert to the user's local time.
    """
    if value is None:
        return None
    if value.tzinfo is None:
        value = value.replace(tzinfo=UTC)
    return int(round(value.timestamp() * 1000))


Model = db.Model
if TYPE_CHECKING:
    from app.helpers.db_model_base import DbModelBase
    from app.models import Household, Recipe, User
    from app.models.agent_persona import AgentPersona
    from app.models.agent_recipe_card import AgentRecipeCard

    Model = DbModelBase


class AgentMessageRole(enum.Enum):
    SYSTEM = "system"
    USER = "user"
    ASSISTANT = "assistant"
    TOOL = "tool"


class AgentChat(Model):
    __tablename__ = "agent_chat"

    id: Mapped[int] = db.Column(db.Integer, primary_key=True)
    household_id: Mapped[int] = db.Column(
        db.Integer, db.ForeignKey("household.id"), nullable=False, index=True
    )
    user_id: Mapped[int] = db.Column(
        db.Integer, db.ForeignKey("user.id"), nullable=False, index=True
    )
    title: Mapped[str | None] = db.Column(db.String(255))
    # ``True`` when the title was set explicitly by the user; auto-rename
    # never touches a locked title.
    title_locked: Mapped[bool] = db.Column(db.Boolean(), nullable=False, default=False)
    # ``True`` while the title is still the auto-generated truncate fallback
    # so the LLM-based rename can refine it once enough context exists.
    title_auto: Mapped[bool] = db.Column(db.Boolean(), nullable=False, default=True)
    persona_id: Mapped[int | None] = db.Column(
        db.Integer,
        db.ForeignKey("agent_persona.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
    )

    household: Mapped[Household] = cast(
        Mapped["Household"],
        db.relationship("Household", uselist=False),
    )
    user: Mapped[User] = cast(
        Mapped["User"],
        db.relationship("User", uselist=False),
    )
    persona: Mapped[AgentPersona | None] = cast(
        Mapped["AgentPersona | None"],
        db.relationship("AgentPersona", uselist=False),
    )
    messages: Mapped[list[AgentMessage]] = cast(
        Mapped[list["AgentMessage"]],
        db.relationship(
            "AgentMessage",
            back_populates="chat",
            cascade="all, delete-orphan",
            order_by="AgentMessage.id",
        ),
    )
    cards: Mapped[list[AgentRecipeCard]] = cast(
        Mapped[list["AgentRecipeCard"]],
        db.relationship(
            "AgentRecipeCard",
            back_populates="chat",
            cascade="all, delete-orphan",
            order_by="AgentRecipeCard.position, AgentRecipeCard.id",
        ),
    )

    @classmethod
    def find_for_user(cls, household_id: int, user_id: int) -> list[Self]:
        return (
            cls.query.filter(cls.household_id == household_id, cls.user_id == user_id)
            .order_by(cls.updated_at.desc())
            .all()
        )

    @classmethod
    def find_for_user_with_summary(
        cls, household_id: int, user_id: int
    ) -> list[tuple[Self, int, str | None, datetime | None]]:
        """Return chats with their message count, last user message and the
        timestamp of their most recent message in one query.

        This avoids the N+1 problem ``obj_to_dict`` would otherwise trigger
        when listing many chats: rather than loading every message of every
        chat to compute ``message_count`` / ``last_user_message`` in Python,
        the same values are computed via correlated subqueries.

        Sorting and the displayed timestamp are based on the timestamp of
        the most recent message, NOT the chat row's ``updated_at`` (which
        also bumps on metadata changes like rename / persona switch and
        would otherwise make every chat appear "modified" simultaneously).
        """
        msg_count = (
            select(func.count(AgentMessage.id))
            .where(AgentMessage.chat_id == cls.id)
            .correlate(cls)
            .scalar_subquery()
        )
        last_user = (
            select(AgentMessage.content)
            .where(
                AgentMessage.chat_id == cls.id,
                AgentMessage.role == AgentMessageRole.USER,
            )
            .order_by(AgentMessage.id.desc())
            .limit(1)
            .correlate(cls)
            .scalar_subquery()
        )
        last_msg_at = (
            select(func.max(AgentMessage.updated_at))
            .where(AgentMessage.chat_id == cls.id)
            .correlate(cls)
            .scalar_subquery()
        )
        stmt = (
            select(cls, msg_count, last_user, last_msg_at)
            .where(cls.household_id == household_id, cls.user_id == user_id)
            .order_by(func.coalesce(last_msg_at, cls.updated_at).desc())
        )
        return [
            (row[0], row[1] or 0, row[2], row[3])
            for row in db.session.execute(stmt).all()
        ]

    def obj_to_dict(
        self,
        skip_columns: list[str] | None = None,
        include_columns: list[str] | None = None,
    ) -> dict[str, Any]:
        # Lightweight view: avoid loading the full ``messages`` collection.
        # Use :meth:`obj_to_summary_dict` when ``message_count`` /
        # ``last_user_message`` are needed (preferred path: pre-computed via
        # :meth:`find_for_user_with_summary` to keep the query count flat).
        return super().obj_to_dict(skip_columns, include_columns)

    def obj_to_summary_dict(
        self,
        message_count: int,
        last_user_message: str | None,
        last_message_at: datetime | None = None,
    ) -> dict[str, Any]:
        res = self.obj_to_dict()
        res["message_count"] = message_count
        res["last_user_message"] = last_user_message
        # Emit as the same UTC-millisecond integer the rest of the API uses
        # (see ``KitchenOwlJSONProvider``). This keeps the frontend tz-aware
        # because a naive ``isoformat()`` string would otherwise be parsed
        # as local time on the client.
        res["last_message_at"] = _to_utc_ms(last_message_at)
        return res

    def obj_to_full_dict(self) -> dict[str, Any]:
        res = self.obj_to_dict()
        msgs = self.messages
        res["messages"] = [m.obj_to_dict() for m in msgs]
        last_user = next(
            (m for m in reversed(msgs) if m.role == AgentMessageRole.USER), None
        )
        res["message_count"] = len(msgs)
        res["last_user_message"] = last_user.content if last_user else None
        last_msg = msgs[-1] if msgs else None
        res["last_message_at"] = _to_utc_ms(last_msg.updated_at if last_msg else None)
        # Open right-side recipe cards. Imported lazily to avoid a circular
        # import at module load time.
        from app.models.agent_recipe_card import AgentRecipeCard

        res["cards"] = [
            c.obj_to_dict() for c in AgentRecipeCard.find_open_for_chat(self.id)
        ]
        return res


class AgentMessage(Model):
    __tablename__ = "agent_message"

    id: Mapped[int] = db.Column(db.Integer, primary_key=True)
    chat_id: Mapped[int] = db.Column(
        db.Integer, db.ForeignKey("agent_chat.id"), nullable=False, index=True
    )
    role: Mapped[AgentMessageRole] = db.Column(
        db.Enum(AgentMessageRole), nullable=False
    )
    content: Mapped[str | None] = db.Column(db.Text())
    # JSON-encoded list of tool calls when role == ASSISTANT and the model
    # decided to use tools. Persisted as text for portability across DBs.
    tool_calls: Mapped[str | None] = db.Column(db.Text())
    # OpenAI tool_call_id this message responds to, when role == TOOL.
    tool_call_id: Mapped[str | None] = db.Column(db.String(128))
    tool_name: Mapped[str | None] = db.Column(db.String(128))
    requires_confirmation: Mapped[bool] = db.Column(
        db.Boolean(), nullable=False, default=False
    )
    # Set when this assistant/tool turn produced a recipe so the UI can show
    # a "view recipe" affordance.
    created_recipe_id: Mapped[int | None] = db.Column(
        db.Integer, db.ForeignKey("recipe.id", ondelete="SET NULL")
    )
    # JSON-encoded list of inverse operations needed to undo the mutation
    # this tool message performed. Populated by ``RecipeAgent`` for tool
    # messages whose handler mutated KitchenOwl data; ``None`` otherwise
    # (read-only tools, errors, assistant/user messages). The shape is
    # ``{"ops": [{"tool": str, "type": "create"|"update"|"delete",
    # "entity": str, "entity_id": int|None, "before": dict|None,
    # "after": dict|None, "extra": dict|None}, ...]}``. The ``before`` /
    # ``after`` snapshots include the entity's ``updated_at`` so the undo
    # service can detect concurrent modifications by other users and
    # refuse to overwrite them.
    undo_snapshot: Mapped[str | None] = db.Column(db.Text())
    # JSON-encoded payload of context the user attached to this message.
    # Shape: ``{"recipe_ids": [int], "item_ids": [int]}``. Only set on
    # USER messages. The agent picks these up when building the prompt and
    # surfaces them as "attached context" to the LLM.
    attachments_json: Mapped[str | None] = db.Column(db.Text())

    chat: Mapped[AgentChat] = cast(
        Mapped["AgentChat"],
        db.relationship("AgentChat", back_populates="messages"),
    )
    created_recipe: Mapped[Recipe | None] = cast(
        Mapped["Recipe | None"],
        db.relationship("Recipe", uselist=False),
    )

    def obj_to_dict(
        self,
        skip_columns: list[str] | None = None,
        include_columns: list[str] | None = None,
    ) -> dict[str, Any]:
        res = super().obj_to_dict(skip_columns, include_columns)
        if isinstance(res.get("role"), AgentMessageRole):
            res["role"] = res["role"].value
        # ``undo_snapshot`` is an internal field used by the rewind/edit
        # endpoints to invert prior tool calls. It can hold large object
        # snapshots and is of no use to the chat UI -- never expose it.
        res.pop("undo_snapshot", None)
        # The UI only needs to know whether this tool call is reversible,
        # not the full snapshot payload. ``has_undo`` is True for any
        # mutating tool call where we captured a snapshot.
        res["has_undo"] = bool(self.undo_snapshot)
        # Decode attachments JSON for the API consumer (avoids forcing the
        # frontend to parse a nested JSON string).
        if self.attachments_json:
            try:
                import json

                res["attachments"] = json.loads(self.attachments_json)
            except Exception:
                res["attachments"] = None
        else:
            res["attachments"] = None
        res.pop("attachments_json", None)
        return res

File: kitchenowl/lib/models/agent_chat.dart
import 'package:equatable/equatable.dart';
import 'package:kitchenowl/models/model.dart';

/// Parses a JSON datetime field that the backend serializes as UTC
/// milliseconds since epoch (see ``KitchenOwlJSONProvider``). Falls back
/// to ``DateTime.tryParse`` when an older backend still sends an ISO
/// string. The result is always tz-aware UTC, so callers can safely
/// ``.toLocal()`` for display.
DateTime? _parseUtcDateTime(dynamic raw) {
  if (raw == null) return null;
  if (raw is int) {
    return DateTime.fromMillisecondsSinceEpoch(raw, isUtc: true);
  }
  if (raw is double) {
    return DateTime.fromMillisecondsSinceEpoch(raw.round(), isUtc: true);
  }
  if (raw is String) {
    final parsed = DateTime.tryParse(raw);
    if (parsed == null) return null;
    // ``DateTime.tryParse`` treats a string without a tz suffix as local
    // time. The server sends UTC, so reinterpret naive results as UTC.
    return parsed.isUtc
        ? parsed
        : DateTime.utc(
            parsed.year,
            parsed.month,
            parsed.day,
            parsed.hour,
            parsed.minute,
            parsed.second,
            parsed.millisecond,
            parsed.microsecond,
          );
  }
  return null;
}

class AgentFileAttachment extends Equatable {
  final String id;
  final String filename;
  final String? mimeType;
  final int? size;
  final DateTime? uploadedAt;

  const AgentFileAttachment({
    required this.id,
    required this.filename,
    this.mimeType,
    this.size,
    this.uploadedAt,
  });

  factory AgentFileAttachment.fromJson(Map<String, dynamic> map) {
    final uploadedAt = _parseUtcDateTime(map['uploaded_at']);

    return AgentFileAttachment(
      id: (map['id'] as String?) ?? '',
      filename: (map['filename'] as String?) ?? (map['id'] as String?) ?? '',
      mimeType: map['mime_type'] as String?,
      size: map['size'] as int?,
      uploadedAt: uploadedAt,
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'filename': filename,
        'mime_type': mimeType,
        'size': size,
        'uploaded_at': uploadedAt?.toIso8601String(),
      };

  @override
  List<Object?> get props => [id, filename, mimeType, size, uploadedAt];
}

class AgentMessageAttachments extends Equatable {
  final List<int> recipeIds;
  final List<int> itemIds;
  final List<AgentFileAttachment> files;

  const AgentMessageAttachments({
    this.recipeIds = const [],
    this.itemIds = const [],
    this.files = const [],
  });

  bool get isEmpty => recipeIds.isEmpty && itemIds.isEmpty && files.isEmpty;

  factory AgentMessageAttachments.fromJson(Map<String, dynamic>? map) {
    if (map == null) return const AgentMessageAttachments();
    final r = map['recipe_ids'];
    final i = map['item_ids'];
    final f = map['files'];
    return AgentMessageAttachments(
      recipeIds: r is List ? r.whereType<int>().toList() : const [],
      itemIds: i is List ? i.whereType<int>().toList() : const [],
      files: f is List
          ? f
              .whereType<Map>()
              .map((e) =>
                  AgentFileAttachment.fromJson(Map<String, dynamic>.from(e)))
              .where((e) => e.id.isNotEmpty)
              .toList()
          : const [],
    );
  }

  Map<String, dynamic> toJson() => {
        'recipe_ids': recipeIds,
        'item_ids': itemIds,
        'files': files.map((f) => f.toJson()).toList(),
      };

  @override
  List<Object?> get props => [recipeIds, itemIds, files];
}

enum AgentMessageRole {
  system,
  user,
  assistant,
  tool;

  static AgentMessageRole fromString(String? value) {
    switch (value) {
      case 'system':
        return AgentMessageRole.system;
      case 'user':
        return AgentMessageRole.user;
      case 'tool':
        return AgentMessageRole.tool;
      case 'assistant':
      default:
        return AgentMessageRole.assistant;
    }
  }
}

class AgentMessage extends Model {
  final int? id;
  final int? chatId;
  final AgentMessageRole role;
  final String? content;
  final String? toolName;
  final String? toolCallId;
  final String? toolCallsJson;
  final int? createdRecipeId;
  final bool hasUndo;
  final bool requiresConfirmation;
  final DateTime? createdAt;
  final AgentMessageAttachments attachments;

  const AgentMessage({
    this.id,
    this.chatId,
    required this.role,
    this.content,
    this.toolName,
    this.toolCallId,
    this.toolCallsJson,
    this.createdRecipeId,
    this.hasUndo = false,
    this.requiresConfirmation = false,
    this.createdAt,
    this.attachments = const AgentMessageAttachments(),
  });

  factory AgentMessage.fromJson(Map<String, dynamic> map) {
    final parsed = _parseUtcDateTime(map['created_at']);
    return AgentMessage(
      id: map['id'],
      chatId: map['chat_id'],
      role: AgentMessageRole.fromString(map['role'] as String?),
      content: map['content'] as String?,
      toolName: map['tool_name'] as String?,
      toolCallId: map['tool_call_id'] as String?,
      toolCallsJson: map['tool_calls'] as String?,
      createdRecipeId: map['created_recipe_id'] as int?,
      hasUndo: map['has_undo'] as bool? ?? false,
      requiresConfirmation: map['requires_confirmation'] as bool? ?? false,
      createdAt: parsed,
      attachments: AgentMessageAttachments.fromJson(
        map['attachments'] is Map
            ? Map<String, dynamic>.from(map['attachments'] as Map)
            : null,
      ),
    );
  }

  @override
  Map<String, dynamic> toJson() => {
        "role": role.name,
        "content": content,
        "tool_name": toolName,
        "tool_call_id": toolCallId,
        "created_recipe_id": createdRecipeId,
      };

  @override
  List<Object?> get props => [
        id,
        chatId,
        role,
        content,
        toolName,
        toolCallId,
        toolCallsJson,
        createdRecipeId,
        hasUndo,
        requiresConfirmation,
        createdAt,
        attachments,
      ];
}

enum AgentRecipeCardSource {
  existing,
  created,
  proposed;

  static AgentRecipeCardSource fromString(String? value) {
    switch (value) {
      case 'created':
        return AgentRecipeCardSource.created;
      case 'proposed':
        return AgentRecipeCardSource.proposed;
      case 'existing':
      default:
        return AgentRecipeCardSource.existing;
    }
  }
}

class AgentRecipeCard extends Model {
  final int id;
  final int chatId;
  final int? recipeId;
  final AgentRecipeCardSource source;
  final String? title;
  final String? description;
  final String? groupLabel;
  final bool closed;

  const AgentRecipeCard({
    required this.id,
    required this.chatId,
    required this.source,
    this.recipeId,
    this.title,
    this.description,
    this.groupLabel,
    this.closed = false,
  });

  factory AgentRecipeCard.fromJson(Map<String, dynamic> map) => AgentRecipeCard(
        id: map['id'] as int,
        chatId: map['chat_id'] as int,
        recipeId: map['recipe_id'] as int?,
        source: AgentRecipeCardSource.fromString(map['source'] as String?),
        title: map['title'] as String?,
        description: map['description'] as String?,
        groupLabel: map['group_label'] as String?,
        closed: map['closed'] as bool? ?? false,
      );

  AgentRecipeCard copyWith({
    String? groupLabel,
    bool clearGroup = false,
  }) =>
      AgentRecipeCard(
        id: id,
        chatId: chatId,
        source: source,
        recipeId: recipeId,
        title: title,
        description: description,
        groupLabel: clearGroup ? null : (groupLabel ?? this.groupLabel),
        closed: closed,
      );

  @override
  Map<String, dynamic> toJson() => {
        'id': id,
        'chat_id': chatId,
        'recipe_id': recipeId,
        'source': source.name,
        'title': title,
        'description': description,
        'group_label': groupLabel,
        'closed': closed,
      };

  @override
  List<Object?> get props =>
      [id, chatId, recipeId, source, title, description, groupLabel, closed];
}

class AgentChat extends Model {
  final int? id;
  final int? householdId;
  final int? userId;
  final String? title;
  final bool titleLocked;
  final bool titleAuto;
  final int? personaId;
  final int messageCount;
  final String? lastUserMessage;
  final DateTime? updatedAt;
  final DateTime? lastMessageAt;
  final List<AgentMessage> messages;
  final List<AgentRecipeCard> cards;

  const AgentChat({
    this.id,
    this.householdId,
    this.userId,
    this.title,
    this.titleLocked = false,
    this.titleAuto = true,
    this.personaId,
    this.messageCount = 0,
    this.lastUserMessage,
    this.updatedAt,
    this.lastMessageAt,
    this.messages = const [],
    this.cards = const [],
  });

  /// Timestamp shown in the chat list. Prefers the timestamp of the most
  /// recent message so renames / persona changes do not visually "bump"
  /// every chat to the same time. Falls back to the chat row's
  /// ``updated_at`` when the server did not provide one (older backend).
  DateTime? get displayTimestamp => lastMessageAt ?? updatedAt;

  factory AgentChat.fromJson(Map<String, dynamic> map) {
    final parsed = _parseUtcDateTime(map['updated_at']);
    final lastMsg = _parseUtcDateTime(map['last_message_at']);

    final rawMsgs = map['messages'];
    final rawCards = map['cards'];
    return AgentChat(
      id: map['id'],
      householdId: map['household_id'],
      userId: map['user_id'],
      title: map['title'] as String?,
      titleLocked: map['title_locked'] as bool? ?? false,
      titleAuto: map['title_auto'] as bool? ?? true,
      personaId: map['persona_id'] as int?,
      messageCount: map['message_count'] as int? ?? 0,
      lastUserMessage: map['last_user_message'] as String?,
      updatedAt: parsed,
      lastMessageAt: lastMsg,
      messages: rawMsgs is List
          ? rawMsgs
              .whereType<Map>()
              .map((e) => AgentMessage.fromJson(Map<String, dynamic>.from(e)))
              .toList()
          : const [],
      cards: rawCards is List
          ? rawCards
              .whereType<Map>()
              .map(
                  (e) => AgentRecipeCard.fromJson(Map<String, dynamic>.from(e)))
              .toList()
          : const [],
    );
  }

  AgentChat copyWith({
    String? title,
    bool? titleLocked,
    bool? titleAuto,
    int? personaId,
    bool clearPersona = false,
  }) =>
      AgentChat(
        id: id,
        householdId: householdId,
        userId: userId,
        title: title ?? this.title,
        titleLocked: titleLocked ?? this.titleLocked,
        titleAuto: titleAuto ?? this.titleAuto,
        personaId: clearPersona ? null : (personaId ?? this.personaId),
        messageCount: messageCount,
        lastUserMessage: lastUserMessage,
        updatedAt: updatedAt,
        lastMessageAt: lastMessageAt,
        messages: messages,
        cards: cards,
      );

  AgentChat withCards(List<AgentRecipeCard> newCards) => AgentChat(
        id: id,
        householdId: householdId,
        userId: userId,
        title: title,
        titleLocked: titleLocked,
        titleAuto: titleAuto,
        personaId: personaId,
        messageCount: messageCount,
        lastUserMessage: lastUserMessage,
        updatedAt: updatedAt,
        lastMessageAt: lastMessageAt,
        messages: messages,
        cards: newCards,
      );

  AgentChat withMessages(List<AgentMessage> newMessages) => AgentChat(
        id: id,
        householdId: householdId,
        userId: userId,
        title: title,
        titleLocked: titleLocked,
        titleAuto: titleAuto,
        personaId: personaId,
        messageCount: newMessages.length,
        lastUserMessage: newMessages
            .lastWhere(
              (m) => m.role == AgentMessageRole.user,
              orElse: () =>
                  const AgentMessage(role: AgentMessageRole.assistant),
            )
            .content,
        updatedAt: updatedAt,
        lastMessageAt: lastMessageAt,
        messages: newMessages,
        cards: cards,
      );

  @override
  Map<String, dynamic> toJson() => {
        "id": id,
        "title": title,
      };

  @override
  List<Object?> get props => [
        id,
        householdId,
        userId,
        title,
        titleLocked,
        titleAuto,
        personaId,
        messageCount,
        lastUserMessage,
        updatedAt,
        lastMessageAt,
        messages,
        cards,
      ];
}
